hashpage.go

```diff
@@ -0,0 +1,340 @@
+	parsed, err = parseTemplate("footer.template", "Transaction")
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	page = append(page, parsed...)
+
+	w.Write(page)
+}
+
+// outputPage formats an outputTransactions struct for use in viewing
+func (es *ExploreServer) outputPage(w http.ResponseWriter, outJSON []byte, oID []byte) {
+	var ot outputTransactions
+	err := json.Unmarshal(outJSON, &ot)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Parse header template
+	page, err := parseTemplate("header.template", fmt.Sprintf("Output %x", oID))
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	parsed, err := es.parseTransactions([]crypto.Hash{ot.OutputTx, ot.InputTx})
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	page = append(page, parsed...)
+
+	// Parse the footer template
+	parsed, err = parseTemplate("footer.template", fmt.Sprintf("Output ID %x", oID))
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	page = append(page, parsed...)
+
+	w.Write(page)
+}
+
+// addressPage formats the list of transactions that an address
+// participated in for human consumption
+func (es *ExploreServer) addressPage(w http.ResponseWriter, addrJSON []byte, address []byte) {
+	var ad addrData
+	err := json.Unmarshal(addrJSON, &ad)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Parse header template
+	page, err := parseTemplate("header.template", fmt.Sprintf("Address %x", address))
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Call the function to parse all of the transactions
+	parsed, err := es.parseTransactions(ad.Txns)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	page = append(page, parsed...)
+
+	// Parse the footer template
+	parsed, err = parseTemplate("footer.template", fmt.Sprintf("Address %x", address))
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	page = append(page, parsed...)
+
+	w.Write(page)
+}
+
+func (es *ExploreServer) contractPage(w http.ResponseWriter, fcJSON []byte, fcid []byte) {
+	var fi fcInfo
+	err := json.Unmarshal(fcJSON, &fi)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Parse header template
+	page, err := parseTemplate("header.template", fmt.Sprintf("File Contract %x", fcid))
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	txns := []crypto.Hash{fi.Contract}
+	txns = append(txns, fi.Revisions...)
+	txns = append(txns, fi.Proof)
+	// Call the parseTransactions function to help with all the transactions
+	parsed, err := es.parseTransactions(txns)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	page = append(page, parsed...)
+
+	// Parse the footer template
+	parsed, err = parseTemplate("footer.template", fmt.Sprintf("File Contract %x", fcid))
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	page = append(page, parsed...)
+
+	w.Write(page)
+}
```

Comment:
because go lacks generics, there's often not much that you can do about repetitive logic